src/main/java/io/strimzi/kafka/quotas/VolumeSourceBuilder.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.kafka.quotas;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.LogDirDescription;
+
+/**
+ * A builder which ensures the <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827 API</a> is available and will throw exceptions if not.
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API">KIP-827 API</a>
+ */
+public class VolumeSourceBuilder implements AutoCloseable {
+
+    private final Supplier<Boolean> kip827Available;
+    private final Function<StaticQuotaConfig.KafkaClientConfig, Admin> adminClientFactory;
+    private Admin adminClient;
+    private StaticQuotaConfig config;
+    private Consumer<Collection<Volume>> volumesConsumer;
+
+    /**
+     * Default production constructor for production usage.
+     * Which will lazily create a Kafka admin client using the supplied config.
+     */
+    @SuppressFBWarnings("MC_OVERRIDABLE_METHOD_CALL_IN_CONSTRUCTOR") //false positive we are just passing the method reference
+    public VolumeSourceBuilder() {
+        this(VolumeSourceBuilder::testForKip827, kafkaClientConfig -> AdminClient.create(kafkaClientConfig.getKafkaClientConfig()));
+    }
+
+    /**
+     * Secondary constructor visible for testing.
+     * @param kip827Available used to determine if KIP-827 API's are available
+     * @param adminClientFactory factory function for creating Admin clients with the builders' config.
+     */
+    /* test */ VolumeSourceBuilder(Supplier<Boolean> kip827Available, Function<StaticQuotaConfig.KafkaClientConfig, Admin> adminClientFactory) {
```

Comment:
I was delaying the evaluation until build time as that would only be executed if the plug-in was actually configured to run and thus the API availability actually matters. 

The delayed execution point is worth documenting, assuming we keep it.

Reply:
Why do you need delayed evaluation though? You're _at most_ loading a class (`LogDirDescription`) and reflecting a single method. That's static information, so the answer is not going to change. And the cost is not _so huge_ that it's worth deferring until you know you need the result.

Reply:
We need delayed execution if we continue to support having the plug-in loaded but disabled as I'd argue if it's disabled it shouldn't explode on with the missing api.

Maybe that's better handled by restructuring the construction logic rather than delaying the test.

Reply:
So I've had a look at restructuring it so the test can be done once and avoid the supplier. The trouble is we construct/inject the `VolumeSourceBuilder` during construction of the `StaticQuotaCallback` which is very helpful for unit testing. 

However that does mean we can't invoke the test during construction as that would prevent the plug-in loading, I can't remember what that does to broker startup, which would be annoying if users have it disabled. I don't know how safe it is to assume that the plug-in is running in a 3.3+ broker (especially as there is no reason it has to run only in Strimzi deployments).

I'm of the opinion its better to delay the execution of the test until its required even if the answer doesn't change (we could use [Guava memoize](https://guava.dev/releases/19.0/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier), iirc guava is on the class path, to make that more explicit but doesn't seem worth it to me) rather than prevent everything loading for older brokers. 

The plug-in could be used without monitoring disk usage to just apply static quotas, so failing construction is probably not the appropriate choice.

Reply:
@robobario and I have gone for a wider ranging refactor than I looked at before and have removed the supplier while preserving the delayed evaluation with 7433cc641160300490017cd06fb3f8a301c57a3c